bin/dpipe.ml

```diff
@@ -34,39 +34,85 @@ let do_command input_size output_size mode level wbits =
+            (fun src -> function
+              | Some max ->
+                  unix_read B.bigstring Unix.stdin src 0 (min max input_size)
+              | None -> unix_read B.bigstring Unix.stdin src 0 input_size )
+            (fun dst len ->
+              let _ = unix_write B.bigstring Unix.stdout dst 0 len in
+              output_size )
+            t
+        in
+        match r with
+        | Ok _ -> ()
+        | Error exn ->
+            Format.eprintf "%a\n%!" Decompress.Gzip_deflate.pp_error exn )
+    | `Zlib -> (
+        let t =
+          Decompress.Zlib_deflate.default ~witness:B.bigstring ~wbits level
+        in
+        let r =
+          Decompress.Zlib_deflate.to_result src dst
+            (fun src -> function
+              | Some max ->
+                  unix_read B.bigstring Unix.stdin src 0 (min max input_size)
+              | None -> unix_read B.bigstring Unix.stdin src 0 input_size )
+            (fun dst len ->
+              let _ = unix_write B.bigstring Unix.stdout dst 0 len in
+              output_size )
+            t
+        in
+        match r with
+        | Ok _ -> ()
+        | Error exn ->
+            Format.eprintf "%a\n%!" Decompress.Zlib_deflate.pp_error exn ) )
   | `Decompression -> (
-      let w = Decompress.Window.create ~witness:B.bigstring in
-      let t = Decompress.Zlib_inflate.default ~witness:B.bigstring w in
-      let r =
-        Decompress.Zlib_inflate.to_result src dst
-          (fun src -> unix_read B.bigstring Unix.stdin src 0 input_size)
-          (fun dst len ->
-            let _ = unix_write B.bigstring Unix.stdout dst 0 len in
-            output_size )
-          t
-      in
-      match r with
-      | Ok _ -> ()
-      | Error exn ->
-          Format.eprintf "%a\n%!" Decompress.Zlib_inflate.pp_error exn )
+    match algo with
+    | `Gzip -> (
+        let w =
+          Decompress.Window.create ~crc:Decompress.Window.crc32
+            ~witness:B.bigstring
+        in
+        let t = Decompress.Gzip_inflate.default ~witness:B.bigstring w in
+        let r =
+          Decompress.Gzip_inflate.to_result src dst
+            (fun src -> unix_read B.bigstring Unix.stdin src 0 input_size)
+            (fun dst len ->
+              let _ = unix_write B.bigstring Unix.stdout dst 0 len in
+              output_size )
+            t
+        in
+        match r with
+        | Ok _ -> ()
+        | Error exn ->
+            Format.eprintf "%a\n%!" Decompress.Gzip_inflate.pp_error exn )
+    | `Zlib -> (
+        let w =
+          Decompress.Window.create ~crc:Decompress.Window.adler32
+            ~witness:B.bigstring
+        in
+        let t = Decompress.Zlib_inflate.default ~witness:B.bigstring w in
+        let r =
+          Decompress.Zlib_inflate.to_result src dst
+            (fun src -> unix_read B.bigstring Unix.stdin src 0 input_size)
+            (fun dst len ->
+              let _ = unix_write B.bigstring Unix.stdout dst 0 len in
```

Comment:
it seems like we're ignoring the error code here and just returning a forced "everything went well" here?

Reply:
Indeed, I wrote it just to show: hey, it's work. But we need to provide good code 👍 !